oracle/pkg/l1Listener/l1Listener.go

```diff
@@ -203,3 +246,76 @@ func (l *L1Listener) watchL1Block(ctx context.Context) error {
 		}
 	}
 }
+
+type RelayQuerier interface {
+	Query(blockNumber int64, blockHash string) (string, error)
+}
+
+type MiniRelayQueryEngine struct {
+	relayUrls []string
+	logger    *slog.Logger
+}
+
+func NewMiniRelayQueryEngine(relayUrls []string, logger *slog.Logger) RelayQuerier {
+	return &MiniRelayQueryEngine{
+		relayUrls: relayUrls,
+		logger:    logger,
+	}
+}
+
+func (m *MiniRelayQueryEngine) Query(blockNumber int64, blockHash string) (string, error) {
+	var wg sync.WaitGroup
```

Comment:
This function should be restructured ideally. The top context should be passed here and you can include the retry logic here itself to simplify the l1 listener loop.

Secondly, if you only care about the first result, you should use errgroup. With errgroup you will get a context cancellation when the first routine exits. You can use this context in the http.RequestWithContext function so that the other requests would be cancelled on the first result.

For the retry logic you can use the backoff library that we added in the new consensus changes (https://github.com/primev/mev-commit/blob/main/cl/redisapp/util/retry.go#L25). Each go routine can retry with backoff till you get the first result.